src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -503,26 +508,34 @@ public void AddError(ProjectId key, DiagnosticData diagnostic)
                 AddError(_projectMap, key, diagnostic);
             }
 
-            private void AddErrors<T>(Dictionary<T, HashSet<DiagnosticData>> map, T key, HashSet<DiagnosticData> diagnostics)
+            private void AddErrors<T>(Dictionary<T, Dictionary<DiagnosticData, int>> map, T key, HashSet<DiagnosticData> diagnostics)
             {
                 var errors = GetErrorSet(map, key);
-                errors.UnionWith(diagnostics);
+                foreach (var diagnostic in diagnostics)
+                {
+                    errors.Add(diagnostic, GetNextIncrement());
+                }
             }
 
-            private void AddError<T>(Dictionary<T, HashSet<DiagnosticData>> map, T key, DiagnosticData diagnostic)
+            private void AddError<T>(Dictionary<T, Dictionary<DiagnosticData, int>> map, T key, DiagnosticData diagnostic)
             {
                 var errors = GetErrorSet(map, key);
-                errors.Add(diagnostic);
+                errors.Add(diagnostic, GetNextIncrement());
+            }
+
+            private int GetNextIncrement()
+            {
+                return _incrementDoNotAccessDirectly++;
```

Comment:
Does this need to be in a lock? Otherwise why the `DoNotAccessDirectly` suffix?

Reply:
to make sure people don't use same increment number twice. if one gets the number through Getnexticrement, they will never get same number twice.